### ISSUE TITLE
M3: New gold-distill step for skypilot/lsf

### ISSUE THREAD

**Issue by cmadam** [cmadam/granite.build#59]
Part of #61.

## Goal

A first-class `gold-distill` step running the `kd-sandbox` GOLD trainer on
`skypilot/lsf/ibm-bluevela`, replacing the `gbansible` role + `launch_enroot.sh` path with a real
granite.build build (build record, artifact, recipe).

Depends on #58 for node count, and on #57 for the `run:` script to actually execute on every node.

## What this step does and does not port

SkyPilot's LSF provisioner already owns `bsub`, `blaunch`, enroot import/flatten, and the topology
export. So `launch_enroot.sh` and `bv_run_kd_distill.sh` are **not ported** — the step replaces only
the *inner* logic of the latter: pick the role, then `accelerate launch`.

Per the agreed design, the container image supplies the pinned deps (`kd-sandbox-distill:0.1.0-uv`,
venv at `/stage/.venv`) and the **trainer comes from the live `/proj` checkout**
(`/proj/granite-build/g4os/kd-sandbox`), which the provisioner already identity-mounts. The `/proj`
dependency is therefore just `gold/*.py` plus `configs/deepspeed/accelerate_deepspeed_zero3.yaml`.
There is no `setup:` phase at all — no clone, no pip.

## Layout

Author at `steps/gold-distill/skypilot/` (`steps/README.md:28`: all new steps live in `steps/`).
`make publish-step` renders to `configurations/assets/environments/skypilot/steps/gold-distill/`;
restrict with `environment_configs.Skypilot.subtypes: [lsf]` and let BlueVela resolve it by
ancestor-walk. Public image, so no `Dockerfile` — model the directory on `steps/byoc/skypilot/`:

```
Makefile  step-template.yaml  src/render_gold_config.py  USAGE.md  README.md
test/test_render_gold_config.py  test/lsf/test_skypilot_lsf_gold_distill.py
test-data/lsf/{build.yaml,buildtest.yaml}
```

## Render the gold config in Python, not a Jinja heredoc

Jinja *does* work in `run:` blocks (`openinstruct-rl/step.yaml:140-180` writes a script that way),
so this is about failure mode, not feasibility. The three rules that silently corrupt a run are
exactly the ones a heredoc handles worst:

- `learning_rate`/`min_lr` must render as `%.2e`. Bare `1e-05` is parsed by PyYAML as a **string**
  and crashes the min_lr handling with a str/float `TypeError`. `yaml.safe_dump(float(f'{lr:.2e}'))`
  makes that structural; a Jinja format filter makes it a convention one edit from breaking.
- Booleans must be lower-case YAML. `safe_dump` does that by construction; Jinja emits `True` unless
  every site remembers `| string | lower`.
- The six online-only keys (`vllm_num_servers`, `top_p`, `use_sampled_opd_loss`,
  `last_message_only`, `clip_alpha`, `opd_importance_sampling`) are emitted **only** when
  `vllm_num_servers > 0`. One testable `if` in Python, versus `{% if %}` inside a quoted heredoc
  inside a YAML literal block — and that is precisely the seam the on-policy phase reopens.

Run it with `/stage/.venv/bin/python` so the config is dumped by the same library `gold.py` parses.
Ship `src/` via `file_mounts: {src: src}`; a relative dest is remapped under `$GB_BUILD_WORKDIR`
(`_build_skypilot_mounts`, `skypilot.py:1651-1665`), which on BlueVela is under `/proj` — identity-
mounted and wrap-exempt.

## step-template.yaml

`type: training`; `outputs.optional.checkpoint.type: model`; `config.gold_config.*` carrying every
hyperparameter from the ansible role's `defaults/main.yml`, notably `use_liger_fused_jsd: false`
(granite's `logits_scaling=10` overflows the fused bf16 JSD kernel -> NaN loss) and
`response_template: "<|im_start|>assistant"` (required for completion masking — the nothink chat
template has no `{% generation %}` tag). Launcher carries the literal `image_id` and `envs` for
`HF_HOME`, `TORCH_NCCL_ENABLE_MONITORING`, `TORCH_NCCL_TIMEOUT_MS`, `NCCL_TIMEOUT`, `USE_LIBUV`,
`PYTORCH_CUDA_ALLOC_CONF`, `DS_SKIP_CUDA_CHECK`, `OMP_NUM_THREADS`.

`NCCL_IB_HCA`/`NCCL_SOCKET_IFNAME` are deliberately omitted: BlueVela's `nccl_tuning_file` is
sourced outside the container before the enroot config is written, and the generated `environ()`
forwards the outer environment. **Verify by inspection at the first 2-node run**
(`grep NCCL_ job.log`); if it does not in fact set them, add both to `envs`.

## run: block

1. `export PATH=/stage/.venv/bin:$PATH`. The dispatcher uses `bash <file>`, non-login, so this
   sticks — a login shell would rerun `/etc/profile` and drop it.
2. Snapshot `RANK`/`TOTAL_NODES`/`NUM_GPUS_PER_NODE`/`MASTER_ADDR`/`MASTER_PORT` into locals, then
   **`unset RANK WORLD_SIZE LOCAL_RANK MASTER_ADDR MASTER_PORT`**. accelerate owns per-process rank;
   the fork exports these as *node-level* inputs. Three existing eval steps already do this defensively.
3. Derive `CONFIG_NAME` from the **allocation** (`_node${NODES}`), never from a build param, so the
   launcher's `_node{N}` bookkeeping cannot disagree with reality.
4. Render the config; rank 0 `cat`s it into the log.
5. `accelerate launch --num_processes $((NODES*GPUS)) --num_machines $NODES --machine_rank
   $NODE_RANK --main_process_ip $MADDR --main_process_port $MPORT --config_file
   $KD_CODE_DIR/configs/deepspeed/accelerate_deepspeed_zero3.yaml $KD_CODE_DIR/gold/gold.py
   --config $CFG --output_dir $CKPT_DIR`, plus `--use_vllm=False` off-policy.
6. **Rank 0 only:** `GB_ARTIFACT_ID:checkpoint GB_ARTIFACT_PATH:${CKPT_DIR}`. Unguarded this fires N
   events, since #57's executor streams every node's output into one driver log.

Checkpoints land under `$GB_BUILD_WORKDIR` (per-build, publishable) rather than the launcher's
hardcoded `$SHARE_PATH/kd-sandbox/checkpoints/`, which ignores `CODE_DIR`.

**On-policy forward compatibility:** because the same script runs on every host with its own
`RANK`/`TOTAL_NODES`, the vLLM/trainer split later becomes a pure step-level branch
(`if NODE_RANK >= NODES - VLLM_N then serve else train`) needing zero further fork changes. That is
why the script reads `TOTAL_NODES` at runtime instead of templating node count in.

## Monitor

Copy the `openinstruct-sft` override shape (`ref: space://monitors/skypilot` +
`log_retrieval.mode: periodic`) — the only proven long-training profile on this env; the default
`on_completion` surfaces nothing until the end. Smoke builds override `poll_interval_seconds` to ~30.
**Pin the `WORKLOAD_STATUS_EVENT` regex against real `job_log` output** from the first smoke run:
`Step:\s\d+` is open-instruct's format, while TRL logs a `{'loss': ...}` dict.

## Verification

`make space` (no `${IMAGE_REF}` residue, valid YAML), `make test`, `make check-published`.

`test/test_render_gold_config.py` is where the expensive bugs get caught cheaply. Minimum cases:
`1e-05` round-trips as a float rendering `1.00e-05` (assert `isinstance(..., float)`);
`gradient_checkpointing` emits `true` not `True`; `vllm_num_servers: 0` omits all six online keys;
`vllm_num_servers >= total_nodes` raises; on-policy with 1 node raises (floor is 2);
`response_template` survives verbatim; `use_liger_fused_jsd` defaults false.
